framework/include/base/ObjectStorage.h

```diff
@@ -0,0 +1,106 @@
+/****************************************************************/
+/*               DO NOT MODIFY THIS HEADER                      */
+/* MOOSE - Multiphysics Object Oriented Simulation Environment  */
+/*                                                              */
+/*           (c) 2010 Battelle Energy Alliance, LLC             */
+/*                   ALL RIGHTS RESERVED                        */
+/*                                                              */
+/*          Prepared by Battelle Energy Alliance, LLC           */
+/*            Under Contract No. DE-AC07-05ID14517              */
+/*            With the U. S. Department of Energy               */
+/*                                                              */
+/*            See COPYRIGHT for full restrictions               */
+/****************************************************************/
+
+#ifndef OBJECTSTORAGE_H
+#define OBJECTSTORAGE_H
+
+// MOOSE includes
+#include "MooseTypes.h"
+
+/**
+ * A general storage container for objects capable of storing threaded copies.
+ */
+template<typename ObjectType>
+class ObjectStorage
```

Comment:
Please see the description in the PR behind the reason of `ObjectStorage`, I haven't used it for its purpose at this point, but will.

`MooseObjectWarehouse` is simply a convenience class for storage of single object types: Dampers, Controls, ...

All warehouses that have multiple objects (AuxKernels, UserObjects, ...) will inherit from WarehouseBase and use a custom set of storage containers, see AuxKernelWarehouse